### PR TITLE
Fix: Transient SignatureDoesNotMatch error from GCS during IT (#4185)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Connected issue: #4185 

### Author

- [x] Target branch is `develop`
- [x] Name of source branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue
- [x] Added `partial` label to PR                                    <sub>or this PR completely resolves the connected issue</sub>

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Branch pushed to GitLab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [ ] Moved connected issue to Merged column
- [ ] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [ ] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [ ] Pushed merge commit to GitLab                                 <sub>or PR is labeled `no sandbox`</sub>
- [ ] Deleted PR branch from GitHub and GitLab
- [ ] Build passes on GitLab<sup>1</sup>

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is* 
pushed determines this checklist item.

### Operator (reindex) 

- [ ] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Review by hannes-ucsc — COMMENTED** [review]
Changes to IT are too invasive. Passing `furl` objects should remain the norm. Just use `Union[furl, str]` where necessary and exploit that `s = str(s)` if `instanceof(s, str)`.

**Review by achave11 — APPROVED** [review]
LGTM, nice work!

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Changes to IT are too invasive. Passing furl objects should remain the norm. Just use `Union[furl, str]` where necessary and exploit that `s == str(s)` if `isinstance(s, str)`.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 1]
```diff
@@ -602,12 +602,12 @@ def update(self,
...
             require(access.headers is None, access.headers)
             signed_url = access.url
-            args = signed_url.args
+            args = furl(signed_url).args
             require('X-Goog-Signature' in args, args)
             self._location = signed_url
 
     @property
-    def location(self) -> Optional[furl]:
+    def location(self) -> Optional[str]:
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:610]: The signature in the base class and all implementations need to be updated.

File: test/integration_test.py [hunk 2]
```diff
@@ -823,13 +823,18 @@ def _test_dos(self, catalog: CatalogName, file_uuid: str, file_ext: str):
             self._assertResponseStatus(response)
             self._validate_file_content(response, file_ext)
 
-    def _get_gs_url_content(self, url: furl, size: Optional[int] = None) -> BytesIO:
+    def _get_gs_url_content(self,
+                            url: Union[furl, str],
+                            size: Optional[int] = None
+                            ) -> BytesIO:
+        if isinstance(url, furl):
```
Comment (hannes-ucsc) [inline test/integration_test.py:830]: This isn't exploiting that `s = str(s)` if `isinstance(s, str)` as I requested.

File: test/service/test_repository_proxy.py [hunk 3]
```diff
@@ -160,7 +160,7 @@ def test_repository_files_proxy(self, mock_get_cached_sources):
                                              'X-Goog-Expires': '900',
                                              'X-Goog-SignedHeaders': 'host',
                                              'X-Goog-Signature': 'SOMESIGNATURE',
-                                         })
+                                         }).url
```
Comment (hannes-ucsc) [inline test/service/test_repository_proxy.py:163]: This is not how we convert furl to str.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Index: test/integration_test.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/integration_test.py b/test/integration_test.py
--- a/test/integration_test.py	(revision ab4d43eec137d55e2139cda1f70a4bf2c1228ef0)
+++ b/test/integration_test.py	(date 1656612415538)
@@ -603,9 +603,10 @@
                            retries: Optional[Union[urllib3.util.retry.Retry, bool, int]] = None,
                            redirect: bool = True,
                            preload_content: bool = True) -> urllib3.HTTPResponse:
-        log.info('GET %s ...', str(url))
+        url = str(url)
+        log.info('GET %s ...', url)
         response = self._http.request('GET',
-                                      str(url),
+                                      url,
                                       retries=retries,
                                       redirect=redirect,
                                       preload_content=preload_content)
```

File: src/azul/plugins/repository/tdr/__init__.py (hunk 1 above)
Comment (danielsotirhos) [reply src/azul/plugins/repository/tdr/__init__.py:610]: Fixed signature in base class. The other implementations in `CannedFileDownload` and `DSSFileDownload` were never updated to `Optional[furl]`.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I get two big fat type warnings in one of the modules modified here. The warnings are justified and caused by these changes.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The fixup commit does not contain all the changes actually made since the last review, Do you know why?

File: test/integration_test.py [hunk 4]
```diff
@@ -823,13 +824,17 @@ def _test_dos(self, catalog: CatalogName, file_uuid: str, file_ext: str):
             self._assertResponseStatus(response)
             self._validate_file_content(response, file_ext)
 
-    def _get_gs_url_content(self, url: furl, size: Optional[int] = None) -> BytesIO:
+    def _get_gs_url_content(self,
+                            url: Union[furl, str],
+                            size: Optional[int] = None
+                            ) -> BytesIO:
+        url = str(url)
```
Comment (hannes-ucsc) [inline test/integration_test.py:831]: This one, for example.

**Comment by danielsotirhos** [comment]
> The fixup commit does not contain all the changes actually made since the last review, Do you know why?

The current fixup commit addresses the 2nd review feedback about the type warnings, along with one additional `furl` to `str` conversion previously missed in `integration_test.py`.

The changes you a referring to were from the 1st review feedback, which were in the fixup commit when 2nd review was requested, and squashed before a 3rd review was requested.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**